src/bridge/GrantsMigration.sol

```diff
@@ -0,0 +1,159 @@
+// SPDX-License-Identifier: BSD-3-Clause-Clear
+
+pragma solidity 0.8.23;
+
+import {Grants} from "../Grants.sol";
+import {NODL} from "../NODL.sol";
+import {BridgeBase} from "./BridgeBase.sol";
+
+contract GrantsMigration is BridgeBase {
+    /// @dev Represents the vesting details of a proposal.
+    struct Proposal {
+        address target; // Address of the grant recipient.
+        uint256 amount; // Total amount of tokens to be vested.
+        Grants.VestingSchedule[] schedules; // Array of vesting schedules.
+    }
+
+    /// @dev Tracks voting and execution status of a proposal.
+    struct ProposalStatus {
+        uint256 lastVote; // Timestamp of the last vote.
+        uint8 totalVotes; // Total number of votes cast.
+        bool executed; // Whether the proposal has been executed.
+    }
+
+    // State variables
+    Grants public immutable grants; // Grants contract.
+    mapping(bytes32 => Proposal) public proposals; // Proposals identified by a hash.
+    mapping(bytes32 => ProposalStatus) public proposalStatus; // Status of each proposal.
+
+    // Events
+    event Granted(bytes32 indexed proposal, address indexed user, uint256 amount, uint256 numOfSchedules);
+
+    /**
+     * @param bridgeOracles Array of addresses authorized to initiate and vote on proposals.
+     * @param token Address of the NODL token used for grants.
+     * @param _grants Address of the Grants contract managing vesting schedules.
+     * @param minVotes Minimum number of votes required to execute a proposal.
+     * @param minDelay Minimum delay before a proposal can be executed.
+     */
+    constructor(address[] memory bridgeOracles, NODL token, Grants _grants, uint8 minVotes, uint256 minDelay)
+        BridgeBase(bridgeOracles, token, minVotes, minDelay)
+    {
+        grants = _grants;
+    }
+
+    /**
+     * @notice Bridges a proposal for grant distribution across chains or domains.
+     * @param paraTxHash Hash of the cross-chain transaction or parameter.
+     * @param user Recipient of the grant.
+     * @param amount Total token amount for the grant.
+     * @param schedules Array of VestingSchedule, detailing the vesting mechanics.
+     */
+    function bridge(bytes32 paraTxHash, address user, uint256 amount, Grants.VestingSchedule[] memory schedules)
+        external
+    {
+        _mustBeAnOracle(msg.sender);
+        _mustNotHaveExecutedYet(paraTxHash);
+
+        if (_proposalExists(paraTxHash)) {
+            _mustNotHaveVotedYet(paraTxHash, msg.sender);
+            _mustNotBeChangingParameters(paraTxHash, user, amount, schedules);
+            _recordVote(paraTxHash, msg.sender);
+        } else {
+            _createProposal(paraTxHash, msg.sender, user, amount, schedules);
+        }
```

Comment:
(general comment, not a blocker) I wonder if some of this logic could be moved to your base contract. It sounds very similar to the logic for liquid tokens. It might be a bit too painful/inelegant to do though

Reply:
It could, I thought about it. I thought about two solutions. However it would involve more intrusive changes to our audited NODLMigration which I wanted to avoid.